src/Esprima/Ast/List.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Esprima.Ast
+{
+    // This structure is like List<> from the BCL except it is designed
+    // to be modifiable by this library during the construction of the AST
+    // but publicly read-only thereafter. The only allocation required on the
+    // heap is the backing array storage for the element. An empty list,
+    // however causes no heap allocation; that is, the array is allocated
+    // on first addition.
+
+    public struct List<T> : IReadOnlyList<T>
+    {
+        T[] _items;
+
+        public List(int capacity)
+        {
+            _items = new T[capacity];
+            Count = 0;
+        }
+
+        public List(List<T> list) : this()
+        {
+            if (list.Count <= 0)
+            {
+                return;
+            }
+
+            _items = new T[list.Count];
+            list._items.CopyTo(_items, 0);
+            Count = list.Count;
+        }
+
+        int Capacity => _items?.Length ?? 0;
+
+        public int Count { get; private set; }
```

Comment:
if count were explicit field could ensure `[MethodImpl(MethodImplOptions.AggressiveInlining)] get => _count;` and of course could directly read/write the field

Reply:
Gosh, is the JIT really that dumb? 😒

Reply:
Heh, maybe I'd say it's constrained, by time and benefit 🙂 [Some seem to think so at least](https://github.com/ravendb/ravendb/blob/d52e7116bd6e191527b7b5ad9781d626429aa6b6/src/Voron/Data/Fixed/FixedSizeTreePage.cs#L37)

Reply:
236f9682fd60e080bccbf41784d362b28b9fe567 should take care of this.